### PR TITLE
[Merged by Bors] - Subscribe to subnets an epoch in advance

### PR DESCRIPTION
**Pull request by pawanjay176**
## Issue Addressed

N/A

## Proposed Changes

Subscibe to subnet an epoch in advance of the attestation slot instead of 4 slots in advance.

### CODE REVIEW

**Comment by pawanjay176** [comment]
@AgeManning This one turned out to be a bit more involved than I expected w.r.t making the tests work.

Subscribing immediately instead of a few slots before the attestations slot leads to the `Subscribe` events for the short-lived and long lived subnets to get queued together. Earlier, the `Subscribe` events for the short lived subnets were received after the required slot delay which we could control.
 
So if we don't want to subscribe to already subscribed subnets, the number of `Subscribe` events we end up getting is not deterministic.
For e.g. suppose we have a short lived subscription for `SubnetId(2)`, we get a new validator subscription request and try to subscribe to `SubnetId(2)` as the long lived subnet for that validator. In this case, we don't  need to add an additional Subscribe event. We just add the subnet to the `random_subnets` hashset so that we don't unsubscribe from it later.

On the plus side, we could remove a bunch of redundant tests where we were checking timing logic for these subscriptions since they are instant now.

**Review by AgeManning — APPROVED** [review]
This looks fine to me. 

Have you tested it out on medalla?

**Comment by pawanjay176** [comment]
> Have you tested it out on medalla?

Yep, working fine for the past 2 days with 8 validators.

**Review by michaelsproul — COMMENTED** [review]
I'm not super familiar with this part of the code, but this looks OK to me.

Just a few minor nitpicks which don't necessarily need fixing before merge

File: beacon_node/network/src/attestation_service/mod.rs [hunk 1]
```diff
@@ -137,6 +129,12 @@ impl<T: BeaconChainTypes> AttestationService<T> {
         }
     }
 
+    /// Return count of all currently subscribed subnets(long-lived **and** short-lived).
+    #[allow(dead_code)]
```
[the remark this answers is not in the document]
Comment (pawanjay176) [reply beacon_node/network/src/attestation_service/mod.rs]: This is what I wanted. Thanks @michaelsproul :)

**Pull request closed without merging**